File: go/tests/realestate_board_test.go
// realestate_board_test.go
//
// Verifies the CircleAI.RealEstate port (realestate_board.go): PropertyKind enum
// ordinals/names, register/list/close, ActiveInSuburb (active + suburb filter,
// newest-first, blank error), and SuburbAverage (exact decimal mean; absent when
// no active listings).

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestRealEstate_EnumOrdinals(t *testing.T) {
	if circleai.PropertyKindApartment != 0 || circleai.PropertyKindLand != 4 {
		t.Fatalf("PropertyKind ordinals wrong")
	}
	if circleai.PropertyKindTownhouse.String() != "Townhouse" {
		t.Fatalf("PropertyKind name wrong")
	}
}

func TestRealEstate_ActiveInSuburbAndAverage(t *testing.T) {
	b := circleai.NewInMemoryRealEstateBoard()
	b.RegisterProperty(circleai.Property{PropertyId: "p1", Suburb: "Sandton", Kind: circleai.PropertyKindApartment, Beds: 2, Baths: 1, FloorAreaM2: 80})
	b.RegisterProperty(circleai.Property{PropertyId: "p2", Suburb: "Sandton", Kind: circleai.PropertyKindHouse, Beds: 4, Baths: 3, FloorAreaM2: 250})
	b.RegisterProperty(circleai.Property{PropertyId: "p3", Suburb: "Rosebank", Kind: circleai.PropertyKindApartment, Beds: 1, Baths: 1, FloorAreaM2: 55})

	t0 := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	b.List(circleai.Listing{ListingId: "L1", PropertyId: "p1", AskingPrice: circleai.DecimalFromInt(1000000), Currency: "ZAR", ListedUtc: t0, IsActive: true})
	b.List(circleai.Listing{ListingId: "L2", PropertyId: "p2", AskingPrice: circleai.DecimalFromInt(3000000), Currency: "ZAR", ListedUtc: t0.Add(24 * time.Hour), IsActive: true})
	b.List(circleai.Listing{ListingId: "L3", PropertyId: "p3", AskingPrice: circleai.DecimalFromInt(900000), Currency: "ZAR", ListedUtc: t0, IsActive: true})

	active, err := b.ActiveInSuburb("sandton") // case-insensitive
	if err != nil {
		t.Fatalf("active in suburb: %v", err)
	}
	// L1 + L2 active in Sandton; newest first -> L2 (t0+24h) then L1.
	if len(active) != 2 || active[0].ListingId != "L2" || active[1].ListingId != "L1" {
		t.Fatalf("active listings wrong: %+v", active)
	}

	// Average of 1,000,000 and 3,000,000 = 2,000,000.
	avg, ok, err := b.SuburbAverage("Sandton")
	if err != nil || !ok || !avg.Equal(circleai.DecimalFromInt(2000000)) {
		t.Fatalf("suburb average = %s ok=%v err=%v, want 2000000", avg, ok, err)
	}

	// Close a listing -> excluded.
	if err := b.Close("L1"); err != nil {
		t.Fatalf("close: %v", err)
	}
	if err := b.Close("ghost"); err == nil {
		t.Fatalf("unknown listing close must error")
	}
	active2, _ := b.ActiveInSuburb("Sandton")
	if len(active2) != 1 || active2[0].ListingId != "L2" {
		t.Fatalf("after close active wrong: %+v", active2)
	}

	// No active listings in a suburb -> (zero, false).
	if _, ok, _ := b.SuburbAverage("Nowhere"); ok {
		t.Fatalf("empty suburb average must be absent")
	}
	if _, err := b.ActiveInSuburb(" "); err == nil {
		t.Fatalf("blank suburb must error")
	}
}

File: typescript/tests/realestate_board.test.ts
// realestate_board.test.ts
// Verifies the CircleAI.RealEstate port: property/listing registration, close,
// active-in-suburb (case-insensitive, listed-desc), suburb average, PropertyKind.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryRealEstateBoard,
  RealEstateDomainContext,
  PropertyKind,
  property,
  listing,
  valuation,
  viewing,
} from "../src/realestate/index";

const D = (s: string) => new Date(s);

describe("PropertyKind enum", () => {
  it("exposes the C# members", () => {
    assert.equal(PropertyKind.Apartment, "Apartment");
    assert.equal(PropertyKind.Land, "Land");
  });
});

describe("InMemoryRealEstateBoard", () => {
  it("lists active properties in a suburb ordered by listed date desc", () => {
    const b = new InMemoryRealEstateBoard();
    b.registerProperty(property("p1", "Sandton", PropertyKind.Apartment, 2, 2, 90));
    b.registerProperty(property("p2", "Sandton", PropertyKind.House, 4, 3, 250));
    b.registerProperty(property("p3", "Rosebank", PropertyKind.Townhouse, 3, 2, 150));
    b.list(listing("l1", "p1", 2_000_000, "ZAR", D("2026-01-01T00:00:00Z"), true));
    b.list(listing("l2", "p2", 6_000_000, "ZAR", D("2026-03-01T00:00:00Z"), true));
    b.list(listing("l3", "p3", 3_000_000, "ZAR", D("2026-02-01T00:00:00Z"), true));
    const sandton = b.activeInSuburb("SANDTON"); // case-insensitive
    assert.deepEqual(
      sandton.map((l) => l.listingId),
      ["l2", "l1"],
    );
    assert.equal(b.suburbAverage("Sandton"), 4_000_000);
  });

  it("close() deactivates and excludes from active + average; empty suburb → null", () => {
    const b = new InMemoryRealEstateBoard();
    b.registerProperty(property("p1", "Sandton", PropertyKind.Apartment, 2, 2, 90));
    b.list(listing("l1", "p1", 2_000_000, "ZAR", D("2026-01-01T00:00:00Z"), true));
    b.close("l1");
    assert.deepEqual(b.activeInSuburb("Sandton"), []);
    assert.equal(b.suburbAverage("Sandton"), null);
    assert.equal(b.suburbAverage("Nowhere"), null);
    assert.throws(() => b.close("ghost"), /Unknown listing ghost/);
    assert.throws(() => b.activeInSuburb(" "), /suburb required/);
  });

  it("records valuations and viewings without error", () => {
    const b = new InMemoryRealEstateBoard();
    assert.doesNotThrow(() => b.value(valuation("p1", 1_800_000, "AVM", D("2026-01-01T00:00:00Z"))));
    assert.doesNotThrow(() => b.scheduleViewing(viewing("v1", "l1", "Buyer", D("2026-01-05T00:00:00Z"))));
    assert.throws(() => b.registerProperty(null as never));
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(RealEstateDomainContext.systemPromptSnippet.includes("[DOMAIN: RealEstate]"));
    assert.deepEqual(RealEstateDomainContext.complianceFlags, [
      "Alienation_of_Land_Act",
      "Rental_Housing_Act",
      "PPRA",
      "FICA",
      "POPIA",
    ]);
    assert.deepEqual(RealEstateDomainContext.suggestedTools, ["property_listings", "document_editor", "map", "analytics"]);
  });
});
